deblur/workflow.py

```diff
@@ -401,12 +396,101 @@ def merge_otu_tables(output_fp, all_tables):
     all_tables: list
         list of filepaths for BIOM tables to merge
     """
+    logger = logging.getLogger(__name__)
+    logger.debug('merge_otu_tables for file %d tables'
+                 ' into file %s' % (len(all_tables), output_fp))
+
     master = load_table(all_tables[0])
     for input_fp in all_tables[1:]:
         master = master.merge(load_table(input_fp))
     write_biom_table(master, output_fp)
 
 
+def get_files_for_table(input_dir,
+                        file_end='.fasta.trim.derep.no_artifacts'
+                        '.msa.deblur.no_chimeras'):
+    """Get a list of files to add to the output table
+
+    Parameters:
+    -----------
+    input_dir : string
+        name of the directory containing the deblurred fasta files
+    file_end : string
+        the ending of all the fasta files to be added to the table
+        (default '.trim.derep.no_artifacts.msa.deblur.no_chimeras')
+
+    Returns
+    -------
+    names : list of tuples of (string,string)
+        list of tuples of:
+            name of fasta files to be added to the biom table
+            sampleid (file names without the file_end and path)
+    """
+    logger = logging.getLogger(__name__)
+    logger.debug('get_files_for_table input dir %s, '
+                 'file-ending %s' % (input_dir, file_end))
+
+    names = []
+    for cfile in listdir(input_dir):
+        cname = join(input_dir, cfile)
+        if not isfile(cname):
+            continue
+        if len(cfile) < len(file_end):
+            continue
+        if cfile[-len(file_end):] == file_end:
+            sampleid = cfile[:-len(file_end)]
+            names.append((cname, sampleid))
+    logger.debug('found %d files' % len(names))
+    return names
+
+
+def create_otu_table(output_fp, deblurred_list):
+    """Create a biom table out of all files in a directory
+
+    Parameters
+    ----------
+    output_fp : string
+        filepath to final BIOM table
+    deblurred_list : list of string
+        list of file names (including path) of all deblurred
+        fasta files to add to the table
+    """
+    logger = logging.getLogger(__name__)
+    logger.debug('create_otu_table for %d samples, '
+                 'into output table %s' % (len(deblurred_list), output_fp))
+
+    seqdict = {}
+    seqlist = []
+    sampdict = {}
+    samplist = []
+    obs = scipy.sparse.dok_matrix((1E9, 1E6), dtype=np.int)
```

Comment:
+1 if easy to have rational bounds go avoid hard coding

Btw, didn't realize you were using scipy dok. Awesome! Thought the earlier
benching was with Python dict.
On Jun 18, 2016 8:05 AM, "Jose Navas" <notifications@github.com> wrote:

> In deblur/workflow.py
> <https://github.com/biocore/deblur/pull/34#discussion_r67602113>:
>
> > +    ----------
> > +    output_fp : string
> > +        filepath to final BIOM table
> > +    deblurred_list : list of string
> > +        list of file names (including path) of all deblurred
> > +        fasta files to add to the table
> > +    """
> > +    logger = logging.getLogger(__name__)
> > +    logger.debug('create_otu_table for %d samples, '
> > +                 'into output table %s' % (len(deblurred_list), output_fp))
> > +
> > +    seqdict = {}
> > +    seqlist = []
> > +    sampdict = {}
> > +    samplist = []
> > +    obs = scipy.sparse.dok_matrix((1E9, 1E6), dtype=np.int)
>
> I see. However, in the case in which this limit is reached, deblur will
> fail and it will be a hard limitation of the software. IMO, this is not
> acceptable. Also, you know the maximum number of samples that you are gonna
> add, its len(deblurred_list), correct? You can then initialize the matrix
> as:
>
> obs = scipy.sparse.dok_matrix((1E9, len(deblurred_list), dtype=np.int)
>
> Then, for overcoming the limitation on the number of observations, you can
> change line 480 to extend the matrix once the limit is reached:
>
> try:
>     obs[cseqid, csampidx] = cfreqexcept IndexError:
>     shape = obs.shape
>     obs.resize((shape[0]*2,  shape[1]))
>     obs[cseqid, csampidx] = cfreq
>
> Note that I'm not changing the shape[1] value because we have already
> initialize it to the correct value. The doubling in size is standard CS
> approach to solve this issue.
>
> This code will not fail even if we have more than 1E9 observations that,
> hey who knows? we may reach at some point, specially if working with really
> diverse environments.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/deblur/pull/34/files/3571ea3b02687547816e6f8122ff2ca3628505d1#r67602113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAc8soAU0--Ti0899hykCf-JNQN-yFp-ks5qNAkbgaJpZM4IxpPJ>
> .
>